examples/engine_headless_ios.rs

```diff
@@ -0,0 +1,117 @@
+//! have a headless mode, but it's not a requirement of the `Engine`
+//! trait (defined in the `wasmer_engine` crate).
+//!
+//! What problem does it solve, and what does it mean?
+//!
+//! Once a Wasm module is compiled into executable code and stored
+//! somewhere (e.g. in memory with the Universal engine, or in a
+//! shared object file with the Dylib engine), the module can be
+//! instantiated and executed. But imagine for a second the following
+//! scenario:
+//!
+//!   * Modules are compiled ahead of time, to be instantiated later
+//!     on.
+//!   * Modules are cross-compiled on a machine ahead of time
+//!     to be run on another machine later one.
+//!
+//! In both scenarios, the environment where the compiled Wasm module
+//! will be executed can be very constrained. For such particular
+//! contexts, Wasmer can be compiled _without_ the compilers, so that
+//! the `wasmer` binary is as small as possible. Indeed, there is no
+//! need for a compiler since the Wasm module is already compiled. All
+//! we need is an engine that _only_ drives the instantiation and
+//! execution of the Wasm module.
+//!
+//! And that, that's a headless engine.
+//!
+//! To achieve such a scenario, a Wasm module must be compiled, then
+//! serialized —for example into a file—, then later, potentially on
+//! another machine, deserialized. The next steps are classical: The
+//! Wasm module is instantiated and executed.
+//!
+//! This example uses a `compiler` because it illustrates the entire
+//! workflow, but keep in mind the compiler isn't required after the
+//! compilation step.
+//!
+//! You can run the example directly by executing in Wasmer root:
+//!
+//! ```shell
+//! cargo run --example engine-headless --release --features "cranelift"
+//! ```
+//!
+//! Ready?
+
+use std::fs::File;
+use std::path::Path;
+use std::str::FromStr;
+use tempfile::NamedTempFile;
+use wasmer::imports;
+use wasmer::wat2wasm;
+use wasmer::Instance;
+use wasmer::Module;
+use wasmer::RuntimeError;
+use wasmer::Store;
+use wasmer::Value;
+use wasmer_compiler::{CpuFeature, Target, Triple};
+use wasmer_compiler_cranelift::Cranelift;
+use wasmer_engine_dylib::Dylib;
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // First step, let's compile the Wasm module and serialize it.
+    // Note: we need a compiler here.
+    let serialized_module_file = {
+        // Let's declare the Wasm module with the text representation.
+        let wasm_bytes = wat2wasm(
+            r#"
+(module
+  (type $sum_t (func (param i32 i32) (result i32)))
+  (func $sum_f (type $sum_t) (param $x i32) (param $y i32) (result i32)
+    local.get $x
+    local.get $y
+    i32.add)
+  (export "sum" (func $sum_f)))
+"#
+            .as_bytes(),
+        )?;
+
+        let compiler_config = Cranelift::default();
+        let triple = Triple::from_str("aarch64-apple-ios")
+            .map_err(|error| RuntimeError::new(error.to_string()))?;
```

Comment:
So this is cross-compilation + headless.

Does it need an example for that? I'm OK but we should rename it to `examples/platform_ios_headless.rs` simply like this I guess. What do you think?

Reply:
It was @syrusakbary suggestion but I think it's quite useful as it gives a place to direct people who query compiling wasm to `.dylib` files for iOS. The current `engine_headless` example assumes that it can deserialize the file again which is a false assumption unless the developer has an M1 Mac (And even then I'm not sure if it works, I don't have one).

I'll rename the example.